kernel: Fix overriding

infinisil opened the pull request:
###### Motivation for this change
This was reported in https://github.com/NixOS/nixpkgs/issues/111504 but further complicated in https://github.com/NixOS/nixpkgs/pull/121518

This should clean it up, though I haven't tested it more than just a basic `linuxPackages.kernel` evaluation, as I'm not familiar with how it's all used.

I'd ideally like <s>@samuela</s> @samueldr, @kira-bruneau, @Atemu and @NeQuissimus to take a look at this, as they've been involved in the above two linked issues

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

infinisil committed:
kernel: Fix overriding

This was reported in https://github.com/NixOS/nixpkgs/issues/111504 but
further complicated in https://github.com/NixOS/nixpkgs/pull/121518

This should clean it up

Infinisil commented on pkgs/os-specific/linux/kernel/generic.nix:
```diff
@@ -204,16 +204,10 @@ let
-      overridableKernel = finalKernel // {
-        override = args:
-          lib.warn (
-            "override is stubbed for NixOS kernel tests, not applying changes these arguments: "
```
But also, I don't know why this is done, there's no reasoning for it in the original PR

Atemu replied:
https://github.com/NixOS/nixpkgs/pull/121518/files#r663978966

Will take a look tomorrow.

samuela commented:
I think you're looking for @samueldr I'm afraid I'm no help here!

Atemu reviewed (changes requested):
`nix-build -A linux.passthru.tests`

```
error: attribute 'override' missing

       at /home/atemu/Projects/nixpkgs/nixos/modules/system/boot/kernel.nix:41:18:

           40|       apply = kernelPackages: kernelPackages.extend (self: super: {
           41|         kernel = super.kernel.override (originalArgs: {
             |                  ^
           42|           inherit randstructSeed;
(use '--show-trace' to show detailed location information)
```

infinisil committed:
kernel: Pass all arguments with callPackage, fixes .override

.override previously didn't exist on kernels because the result from
`callPackage` (which contains a `.override`) was a function which was
then called with some more arguments.

This commit fixes this by passing all arguments in one go via callPackage

Infinisil commented:
@Atemu Ah good catch, I'm surprised ofborg didn't run this. I fixed it with the latest commit.

Atemu commented:
You're a wizard.

ofBorg seems to be catching something else now though...

Infinisil commented:
Closing for now. Hard to get motivation to fix this, it's really hard to wrap your head around this issue
